lgc/patch/VertexFetch.cpp

```diff
@@ -1035,9 +802,183 @@ Value *VertexFetchImpl::fetchVertex(InputImportGenericOp *inst, llvm::Value *des
+  bool is16bitFetch = bitWidth <= 16;
+  bool is64bitFetch = bitWidth == 64;
+  auto fetch64Type = FixedVectorType::get(Type::getInt32Ty(*m_context), 2);
+  auto fetch32Type = FixedVectorType::get(Type::getInt32Ty(*m_context), 4);
+  auto fetch16Type = FixedVectorType::get(Type::getInt16Ty(*m_context), 4);
+  auto fetchType = is16bitFetch ? fetch16Type : fetch32Type;
+
+  Value *args[] = {
+      vbDesc,              // rsrc
+      vbIndex,             // vindex
+      byteOffset,          // offset
+      builder.getInt32(0), // soffset
+      builder.getInt32(0)  // glc, slc
+  };
+
+  Value *wholeVertex = nullptr;
+  {
+    builder.SetInsertPoint(wholeVertexBlock);
+    wholeVertex = builder.CreateIntrinsic(Intrinsic::amdgcn_struct_buffer_load_format, fetchType, args, {});
+    if (is64bitFetch) {
+      // If it is 64-bit, we need the second fetch
+      args[2] = builder.CreateAdd(args[2], builder.getInt32(SizeOfVec4));
+      auto secondFetch = builder.CreateIntrinsic(Intrinsic::amdgcn_struct_buffer_load_format, fetchType, args, {});
+      std::vector<Constant *> shuffleMask;
+      for (unsigned i = 0; i < 8; ++i)
+        shuffleMask.push_back(ConstantInt::get(Type::getInt32Ty(*m_context), i));
+      wholeVertex = builder.CreateShuffleVector(wholeVertex, secondFetch, ConstantVector::get(shuffleMask));
+    }
+    builder.CreateBr(fetchEndBlock);
+  }
+
+  fetchType = FixedVectorType::get(builder.getInt32Ty(), is64bitFetch ? 8 : 4);
+  if (is16bitFetch)
+    fetchType = FixedVectorType::get(builder.getInt16Ty(), 4);
+  Value *lastVert = UndefValue::get(fetchType);
+
+  auto int16Zero = builder.getInt16(0);
+  auto int16One = builder.getInt16(1);
+  auto float16One = builder.getInt16(0x3C00);
+  // Get default fetch values
+  Constant *defaults = nullptr;
+  if (basicTy->isIntegerTy()) {
+    if (bitWidth <= 16)
+      defaults = ConstantVector::get({int16Zero, int16Zero, int16Zero, int16One});
+    else if (bitWidth == 32)
+      defaults = m_fetchDefaults.int32;
+    else {
+      assert(bitWidth == 64);
+      defaults = m_fetchDefaults.int64;
+    }
+  } else if (basicTy->isFloatingPointTy()) {
+    if (bitWidth == 16)
+      defaults = ConstantVector::get({int16Zero, int16Zero, int16Zero, float16One});
+    else if (bitWidth == 32)
+      defaults = m_fetchDefaults.float32;
+    else {
+      assert(bitWidth == 64);
+      defaults = m_fetchDefaults.double64;
+    }
+  } else
+    llvm_unreachable("Should never be called!");
+
+  Value *comp0 = nullptr;
+  Value *comp1 = nullptr;
+  Value *comp2 = nullptr;
+  Value *comp3 = nullptr;
+  auto compType = is16bitFetch ? builder.getInt16Ty() : builder.getInt32Ty();
+
+  // Per-component fetch
+  // X channel
+  // .comp0Block
+  {
+    builder.SetInsertPoint(comp0Block);
+    for (unsigned i = 0; i < fetchType->getNumElements(); i++)
+      lastVert = builder.CreateInsertElement(lastVert, builder.CreateExtractElement(defaults, i), i);
```

Comment:
Isn't this loop just a no-op? (Or at worst incorrect for 64-bit fetches?) Couldn't you just initialize `lastVert` with `defaults`?

Reply:
I'm confused. at that time, I just thought it was constant and couldn't be modified. Thanks.